### ISSUE TITLE
`clang-analyzer-cplusplus.Move` not reported for usage after `std::move()` with range

### ISSUE THREAD

**Issue by firewave** [llvm/llvm-project#137157]
```
#include <algorithm>
#include <iostream>
#include <iterator>
#include <list>
#include <string>

int main()
{
    std::list<std::string> l1 = { "1" };

    std::list<std::string> l2;
    std::move(l1.begin(), l1.end(), std::back_inserter(l2));

    std::cout << "l1: " << *l1.cbegin() << '\n';
    std::cout << "l2: " << *l2.cbegin() << '\n';
}
```
https://godbolt.org/z/Y96a6nv7h

https://en.cppreference.com/w/cpp/algorithm/move

Also not detected by `bugprone-use-after-move` clang-tidy check: #137156

**Comment by steakhal** [comment]
I believe the checker should know and handle the range variant of `std::move`, and it should be a fairly easy bug to fix.
One just needs to debug what controlflow is taken in the checker that makes it miss the bug.

**Comment by steakhal** [comment]
I think what you may see here must be related to invalidation caused by some opaque function call during conservative call evaluation. You can verify this by the following steps:

1) Dump the exploded graph (aka. egraph, the internal representation of the abstract program states) using the `-analyzer-dump-egraph=dump.dot` flag.
2) Once you have the graph dump, it's really inconvenient to look at it by naked eyes. To help here, we the `clang/utils/analyzer/exploded-graph-rewriter.py`, which you can use as: `exploded-graph-rewriter.py dump.dot`. It will produce you a rewritten egraph in an html. Open it in your browser and start tracing the execution paths until the node you believe the use-after-move should be detected.
3) You should check the address (location or MemRegion) that the checker would look at. Then trace backwards in the egraph to see what node introduced it. Have special attention to "Conservative eval call" nodes, as this is likely where it's coming from.
4) Once you have that Node, you can see that the immediate predecessor of that Node should have the right address in the Store/Environment. This should confirm the theory that the address is invalidated by some conservative eval call.

In the static analyzer, whenever you dump the `State->getID()`, that will be a stable identifier that is the "title" of a rewritten exploded node. If you dump the ID of an ExplodedNode, aka. `N->getID()`, that will be in the rewritten egraph too, but in some enumeration within a State.

![Image](https://github.com/user-attachments/assets/eefbeab6-02e5-412d-ba75-a3ace6f2e305)

This will help you correlate between the prints in your terminal and the egraph nodes.

**Comment by hapeeeeee** [comment]
> I think what you may see here must be related to invalidation caused by some opaque function call during conservative call evaluation. You can verify this by the following steps:
> 
> 1. Dump the exploded graph (aka. egraph, the internal representation of the abstract program states) using the `-analyzer-dump-egraph=dump.dot` flag.
> 2. Once you have the graph dump, it's really inconvenient to look at it by naked eyes. To help here, we the `clang/utils/analyzer/exploded-graph-rewriter.py`, which you can use as: `exploded-graph-rewriter.py dump.dot`. It will produce you a rewritten egraph in an html. Open it in your browser and start tracing the execution paths until the node you believe the use-after-move should be detected.
> 3. You should check the address (location or MemRegion) that the checker would look at. Then trace backwards in the egraph to see what node introduced it. Have special attention to "Conservative eval call" nodes, as this is likely where it's coming from.
> 4. Once you have that Node, you can see that the immediate predecessor of that Node should have the right address in the Store/Environment. This should confirm the theory that the address is invalidated by some conservative eval call.
> 
> In the static analyzer, whenever you dump the `State->getID()`, that will be a stable identifier that is the "title" of a rewritten exploded node. If you dump the ID of an ExplodedNode, aka. `N->getID()`, that will be in the rewritten egraph too, but in some enumeration within a State.

> This will help you correlate between the prints in your terminal and the egraph nodes.

In the `modelUse` function, I used the approach below to confirm that two different `MemRegion`s actually originated from the same `vector[idx]` expression:

```c++
void MoveChecker::modelUse(ProgramStateRef State, const MemRegion *Region,
                           const CXXRecordDecl *RD, MisuseKind MK,
                           CheckerContext &C) const {
  assert(!C.isDifferent() && "No transitions should have been made by now");
  const RegionState *RS = State->get<TrackedRegionMap>(Region);


  if (const auto *SR = dyn_cast<SymbolicRegion>(Region)) {
    if (const auto *SCE = dyn_cast<SymbolConjured>(SR->getSymbol())) {
      const Stmt *StmtCtx = SCE->getStmt();
      printf("Stmt class name: %s\n", StmtCtx->getStmtClassName()); // Stmt class name: CXXOperatorCallExpr

      if (const auto *COE = dyn_cast<CXXOperatorCallExpr>(StmtCtx)) {
        const Expr *VecExpr = COE->getArg(0);
        const Expr *IdxExpr = COE->getArg(1);
        printf("Operator: %s\n", COE->getDirectCallee()->getNameAsString().c_str()); // Operator: operator[]
        
        if (const auto *DRE = dyn_cast<DeclRefExpr>(VecExpr->IgnoreParenCasts())) {
          if (const auto *VD = dyn_cast<VarDecl>(DRE->getDecl())) {
            printf("Vector var: %s(%p)\n", VD->getNameAsString().c_str(), VD); // same pointer for VD
          }
        }
      }
    }
  }
  
...
```

Once I get the `VarDecl`, I can retrieve the `ElementRegion` for the specific element using:

```c++
const VarRegion *VarReg = MRMgr.getVarRegion(VD, C.getLocationContext());
const ElementRegion *ElemReg = MRMgr.getElementRegion(ElementTy, IndexVal, VarReg, C.getASTContext());
```

This helps address the issue in my specific test case.
However, this approach is not very general — it only works for `CXXOperatorCallExpr`, and only for types compatible with `getElementRegion`.

What I'm looking for is a more generic way to recover the actual `ElementRegion` corresponding to `v[1]`, starting from a `SymbolicRegion` (created by a `ConjuredSymbol`).
I’ve tried `getBaseRegion()` and `getOriginalRegion()` on the `SymbolicRegion`, but they don't seem to give me back the original vector region.

Any suggestions on how to extract the actual memory region of `v[1]` in a general way from the symbolic region would be greatly appreciated!

**Comment by steakhal** [comment]
I can only recommend looking at the exploded graph. There, you should have the give conjured symbol spelled as `conj_$NNN`.
From just looking at the AST it's not easy (actually impossible) to recover the exact index. That is only present in the exploded graph.

**Comment by steakhal** [comment]
If you dump the ID of the State ID, you can quickly related at that point you are in the exploded graph within this `MoveChecker::modelUse` call. Then you can access the values from the Environment of the State using the `C.getSVal(expr).dump()`.

**Comment by steakhal** [comment]
> Hi @steakhal. I have taken a look at the exploded graph as you suggested in your comment. When I looked at it I could see that the analyzer inlined `std::move(first, list, d_first)` into STL internals. The `cplusplus.Move` checker marks the internal iterator `(__it)` as moved, but not any elements of l1. The move per element `(*dest = std::move(*src))` happens inside `std::__copy_move_a1` and it does not get inlined. Because of that the checker does not have a chance to observe the move constructors on l1 elements.

Ideally, we would not inline the 3 param std::move, but have a model for it. And the move checker would do the right thing.

**Comment by benedekaibas** [comment]
I have added evalCall to MoveChecker to mark the moved state. The key point of my implementation is to recover the source container from the iterator arguments. I handle the cases currently where the first argument is a begin() member call on the source container. 

```text
move_iterator.cpp:14:29: warning: Method called on moved-from object 'l1' of type 'std::list' [cplusplus.Move]
   14 |     std::cout << "l1: " << *l1.cbegin() << '\n'; // <--- should have a use-after-move
```

From where the call happened I walk back to the container object and obtain its region through `getLValue` since `getSVal` returns unknown at this point of the analysis (I have tried to use that first, but it did not work for detecting the error). I have reused the existing classifyObject and shouldBeTracked checks logic and mark it in `TrackedRegionMap` as moved. Then `checkPreCall` handles the warning when the container is used afterwards. My current implementation tracks the container region as moved and not the individual elements. Because of this I currently have false positives on methods like `size()` when I tested my implementation on some code examples. 

```text
move_iterator_three.cpp:17:18: warning: Method called on moved-from object 'l1' of type 'std::list' [cplusplus.Move]
   17 |     std::cout << l1.size();
```

This is where I am struggling right now. I have an idea of tracking these cases with a new state in `RegionState`, something like `ContentsMoved` to distinguish from a full container move. But I am wondering what your opinion is about this implementation plan?

**Comment by benedekaibas** [comment]
> This is where I am struggling right now. I have an idea of tracking these cases with a new state in `RegionState`, something like `ContentsMoved` to distinguish from a full container move.

Hi @steakhal 

What I have done to resolve this issue is that I have added another program state map  to distinguish from the existing program state map.

```
REGISTER_MAP_WITH_PROGRAMSTATE(TrackedRegionMap, const MemRegion *, RegionState)
REGISTER_MAP_WITH_PROGRAMSTATE(TrackedContentsMap, const MemRegion *, RegionState)
```

In `evalCall` when the 3-arg `std::move` matches I mark the source container in `TrackedContentsMap`. When I used `TrackedRegionMap` the checker did report emit an error message on the code example that is in this issue, but it also did report on any method call on the container including safe method calls like `size()`. I have figured this out after testing my original implementation (the one that included `TrackedRegionMap` in `evalCall`) not just on the original examples, but some other examples that I built to see if my implementation crashes on them. That's the reason why I have added another program state map. To catch the original error and not producing false positives in `checkPreCall` I detect calls to `operator*` and `operator->` then I recover the iterator's container through `getIteratorPosition()` and check whether that container is in `TrackedContentsMap`. If yes then I emit a warning through `modelUse()`. The problem is that `modelUse()` currently works with `TrackedRegionMap` and not with the program state map I have added, so when I try to use it to emit the report for the unsafe method calls because it cannot look up `TrackedContentsMap` I decided to use `tryReportBug()` instead.

```cpp
    if(State->get<TrackedContentsMap>(ContainerRegion)) {
      ExplodedNode *N = tryToReportBug(ContainerRegion, R, C, MK_FunCall);
      if (!N || N->isSink())
        return;

      State = State->set<TrackedContentsMap>(ContainerRegion, RegionState::getReported());
      C.addTransition(State, N);
    }
```

For this implementation I modeled this after the pattern used in the `modelUse` and reused it in for the detection of the `operator*` and `operator->` that lives in `checkPreCall`.

I thought it would be helpful to explain the steps I have done to solve the issue and how I have implemented the theories I have formed in my mind. I wanted to discuss my approach first to see if there is a better alternative that you would suppose me to implement instead of the implementation I have come up with.

Thank you for your time and help!

**Comment by steakhal** [comment]
I just realised that I still owe an answer here. Sorry about that.

> > This is where I am struggling right now. I have an idea of tracking these cases with a new state in `RegionState`, something like `ContentsMoved` to distinguish from a full container move.
> 
> Hi steakhal
> 
> What I have done to resolve this issue is that I have added another program state map to distinguish from the existing program state map.
> [...]
> I thought it would be helpful to explain the steps I have done to solve the issue and how I have implemented the theories I have formed in my mind. I wanted to discuss my approach first to see if there is a better alternative that you would suppose me to implement instead of the implementation I have come up with.
>
> Thank you for your time and help!

I think the approach seems reasonable to me. Iterator modeling/container modeling is tricky because sometimes they are objects (at time rvalues, in other words symbols) and other times they are memregions, such as a pointer.

I want to be sure I understand your question. What two situation can't your checker distinguish?

**Comment by benedekaibas** [comment]
> I want to be sure I understand your question. What two situation can't your checker distinguish?


So my original solution was the following logic in my `evalCall()` function:

```cpp
  const MemRegion *Region = State->getLValue(VD, C.getLocationContext()).getAsRegion();
  const CXXRecordDecl *RD = ContainerExpr->getType()->getAsCXXRecordDecl();

  // For both Region and RD I check if they are null -> return false

  ObjectKind OK = classifyObject(State, Region, RD);

  if (shouldBeTracked(OK)) {
    State = State->set<TrackedRegionMap>(Region, RegionState::getMoved());
  }
  C.addTransition(State);
  return true;
```

So I get the declaration of the region of the container as `LVal` and then I also record the region where the container is declared. I check whether the container should be tracked and if yes then I mark it as moved. This was my original implementation and it detected the following example:

`std::cout << "l1: " << *l1.cbegin() << '\n'; // <--- should have a use-after-move`

```text
move_iterator.cpp:14:28: warning: Method called on moved-from object 'l1' of
      type 'std::list' [cplusplus.Move]
   14 |     std::cout << "l1: " << *l1.cbegin() << '\n'; // <--- should ...
      |                            ^~~~~~~~~~~~
```

The problem was that when I tested this implementation on different examples, like:

```cpp
  std::move(myList.begin(), myList.end(), std::back_inserter(l2));
  std::cout << myList.size();
```

My implementation caused a false positive for the safe `size()` method call on the container. Since I track the container any method call on the container would lead for the checker to emit a warning even the safe method calls like `size` in this example. 

So at this implementation my checker could not distinguish between full container moved and contents of the container moved cases. I think this is what your question refers to, but feel free to correct me if I misunderstand something. 

That is the reason why I have come up with the idea of having another program state map. For that I could just update my implementation in `evalCall` to use the new program state map:

```cpp
  const MemRegion *Region = State->getLValue(VD, C.getLocationContext()).getAsRegion();
  const CXXRecordDecl *RD = ContainerExpr->getType()->getAsCXXRecordDecl();

  // For both Region and RD I check if they are null -> return false

  ObjectKind OK = classifyObject(State, Region, RD);

  if (shouldBeTracked(OK)) {
    State = State->set<TrackedContentsMap>(Region, RegionState::getMoved());
  }
  C.addTransition(State);
  return true;
```

Then in `checkPreCall` I detect "unsafe" calls like `operator*`, `operator->` then I recover the iterator's container through `getIteratorPosition()` and check whether that container is in `TrackedContentsMap`. If yes then I emit a warning through `tryToReportBug`.

I handle safe calls like `size` by allowing the call to proceed without warning since the container is only in `TrackedContentsMap` and not in `TrackedRegionMap` so the exising warning won't warn either.

**Comment by steakhal** [comment]
> So at this implementation my checker could not distinguish between full container moved and contents of the container moved cases. I think this is what your question refers to, but feel free to correct me if I misunderstand something.

You are right. To be more accurate, we could distinguish these because the container and the elements of the container have different memory regions, but the problem is that we can't enumerate all of the moved elements because there are symbolic many of them (`size` many). And this leaves the only option to special-case this scenario.
Another problem is that unlike with the 1 param `std::move`, this ranged `move` does not operate directly on the contrainer but via an indirection: and iterator to the container.
This means that to know at the point of `std::move` we need to know what the iterator points to, thus there is a natural dependency on the IteratorModeling.
For now, I think it's fine to say that the range-based `move` would only work if `IteratorModeling` was enabled. It just shouldn't crash if it's not enabled.
